x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.qa.single_node;
+
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.elasticsearch.common.CheckedConsumer;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.xpack.sql.qa.jdbc.JdbcIntegrationTestCase;
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static java.lang.String.join;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.common.collect.Tuple.tuple;
+import static org.hamcrest.collection.IsEmptyCollection.empty;
+
+/**
+ * <p>This test was introduced because of the inconsistencies regarding NULL argument handling. NULL literal vs NULL field 
+ * value as function arguments in some case result in different function return values.</p>
+ *
+ * <p>Functions should return with the same value no matter if the argument(s) came from a field or from a literal.</p>
+ * 
+ * <p>The test class based on the example function calls (and argument specifications) generates all the 
+ * permutations of the function arguments (4 options per argument: value/NULL as literal/field) and tests that the 
+ * function calls with the same argument values provide the same result regardless of the source (literal, field) 
+ * of the arguments.</p>
+ *  
+ * <p>To ignore any of the tests, add an .ignore() method call after the Fn ctors in the FUNCTION_CALLS_TO_TEST list below, like:
+ * <code> new Fn("ASCII", "foobar").ignore()</code></p>
+ */
+public class ConsistentFunctionArgHandlingIT extends JdbcIntegrationTestCase {
```

Comment:
+1 for generating the test case to test various combinations however I find the current approach too convoluted.
Instead of indexing the data, filtering the functions and generating queries through an integration test one could use the `FunctionRegistry` directly and just focus on the string functions either by declaring their arguments as in this test or inspecting the returned definitions.

Specify the arguments or generate the loops based on the signature and pass them to the definition to get back the actual function instance to fold.

The resulting code should be smaller, without strings and faster due to be being a unit test.

Reply:
I agree that writing a unit test  would result in faster tests, however testing if:

1. `processors`
2. `Function.fold()` (optimizer folding rule that calls that if all args are foldable, calls the processor)
3. `optimizer rules` that rewrite the function call in the query (NULL folding for example that depends on `Function.nullable()`)
4. `QueryTranslator` output (function calls turn into painless scripts that call the processors within Query DSL queries)

all work together in a consistent way providing the same outcome regardless the argument sources being field/literal is easier through integration tests, because  the different argument source combinations will hit all of the above code paths using the same test code.

I agree though that in unit tests like [LocateProcessorTests](https://github.com/elastic/elasticsearch/blob/baa147cb7c5a39954c279d0876c49ad32f3d8312/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/LocateProcessorTests.java#L49) (or one similar to this in the PR) we could easily test if `fold()` and the `processor` leads to the same results, and likely it'll catch some of the bugs. But it will not cover `#3` and `#4`.

To cover `#3` we would have to make sure we call the `Optimizer` too, check if the function transformed by the rules will provide an output equivalent to `fold` and `processor` calls.
To cover `#4` we would have to execute the query generated by the `QueryTranslator` (or at least the generated Painless script), which I think is easier to do through integration tests and would require quite a bit of mocking in a unit test.

This looks too much of a whitebox testing for me. I'd rather execute the same code that the user's code would execute with less assumptions about our implementation. IMO it is easier to argue about the coverage of this integration test that uses the same approach to hit all the four cases above.

Discovery: `SqlFunctionRegistry` is not accessible in integration test (not impossible to fix), but even if it would be (or would want to use in a unit test): I don't see how can I easily figure out the type of the arguments from the `FunctionDefinition`. The constructor arguments of the functions are `Expression`s without hints about the actual accepted types and the `Builder`s also don't provide help. One way is to construct a function with generic/widely unacceptable arguments, call `resolveType()` and parse the error messages. It would help if the functions would be self-decribing (we could even generate  a function arguments meta table with `type`, `optional`, `accepted sources` per function that might be useful somewhere else), but it is outside of the scope of this PR.